op-e2e/actions/interop/interop_fork_test.go

```diff
@@ -0,0 +1,452 @@
+		c.Sequencer.SyncSupervisor(t) // supervisor to react to exhaust-L1
+		svr.ProcessFull(t)
+		c.Sequencer.ActL2PipelineFull(t) // node to complete syncing to L1 head.
+
+		// Verify
+		expStatus.SafeL2 = expStatus.UnsafeL2
+		assertHeadBlockIDs(t, c, expStatus)
+		assertSupervisorChainHeads(t, svr, c, expStatus.UnsafeL2.ID(), expStatus.SafeL2.ID(), expStatus.FinalizedL2.ID())
+	}
+
+	// Verify proofs agree
+	assertProgramOutputMatchesDerivationForBlockTimestamp(gt, system, system.Actors.ChainA.Sequencer.L2Unsafe().Time)
+
+	// Advance L1 safe head and finalized head
+	system.Actors.L1Miner.ActL1SafeNext(t)
+	system.Actors.L1Miner.ActL1FinalizeNext(t)
+	system.Actors.L1Miner.ActL1SafeNext(t)
+	system.Actors.L1Miner.ActL1FinalizeNext(t)
+
+	// Sync supervisor
+	svr.SignalLatestL1(t) // supervisor will be aware of latest L1
+	svr.SignalFinalizedL1(t)
+	svr.ProcessFull(t)
+
+	assertL1Heads(t, system, 3, 3)
+
+	for _, c := range chains {
+		expStatus := expSyncStatuses[c]
+
+		// Process L1 finalization
+		c.Sequencer.SyncSupervisor(t)
+		c.Sequencer.ActL2PipelineFull(t)
+		c.Sequencer.SyncSupervisor(t)
+		c.Sequencer.ActL2PipelineFull(t)
+
+		expStatus.FinalizedL2 = expStatus.SafeL2
+		updateExpectedUnsafe(t, c, expStatus, 2) // Assert unsafe head does not move
+		assertHeadBlockIDs(t, c, expStatus)
+	}
+
+	// Verify proofs agree on prev blocks
+	assertProgramOutputMatchesDerivationForBlockTimestamp(gt, system, system.Actors.ChainA.Sequencer.L2Safe().Time)
+
+	for _, c := range chains {
+		expStatus := expSyncStatuses[c]
+		// Advance unsafe head again, going to repeat the above for the second interop block, except this time we'll promote localSafe at the same time as crossUnsafe
+		c.Sequencer.ActL2EmptyBlock(t)
+
+		// Verify
+		assertUnsafeTimeOffset(t, c, genesisTime, 6)
+		updateExpectedUnsafe(t, c, expStatus, 3)
+		assertHeadBlockIDs(t, c, expStatus)
+		// NB: Intentionally asserting that the supervisor's view of the unsafe head is the same as the sequencer's view of the safe head
+		assertSupervisorChainHeads(t, svr, c, expStatus.SafeL2.ID(), expStatus.SafeL2.ID(), expStatus.FinalizedL2.ID())
+	}
+
+	// Settle both chains to DA
+	system.Actors.ActBatchAndMine(t, chains...)
+	assertL1Heads(t, system, 4, 3)
+
+	for _, c := range chains {
+		expStatus := expSyncStatuses[c]
+
+		// Sanity check nothing's changed
+		assertHeadBlockIDs(t, c, expStatus)
+		// NB: Intentionally asserting that the supervisor's view of the unsafe head is the same as the sequencer's view of the safe head
+		assertSupervisorChainHeads(t, svr, c, expStatus.SafeL2.ID(), expStatus.SafeL2.ID(), expStatus.FinalizedL2.ID())
+
+		// Promote localSafe and crossUnsafe
+		svr.SignalLatestL1(t)
+		c.Sequencer.SyncSupervisor(t) // supervisor to react to exhaust-L1
+		svr.ProcessFull(t)
+		c.Sequencer.ActL2PipelineFull(t) // node to complete syncing to L1 head.
+
+		// Verify
+		expStatus.LocalSafeL2 = expStatus.UnsafeL2
+		expStatus.CrossUnsafeL2 = expStatus.UnsafeL2
```

Comment:
Q: Why is it that crossUnsafe and localSafe can be equal to each other and yet not equal to safe? i.e. if the following line were inserted here the test fails on the next call to assertHeadBlockIDs()

    expStatus.SafeL2 = expStatus.UnsafeL2

Reply:
I think this can happen if the remote chains haven't been fully derived yet, but their unsafe heads are known. So then the local unsafe == cross unsafe, and because we haven't generated more unsafe blocks and batched and derived the latest unsafe chain, local safe (derived) == unsafe.